design/parallel-commit/parallel-commit-solution-ideas.md

```diff
@@ -0,0 +1,59 @@
+# Parallel Commit Solution Ideas
+
+We have many difficulties to overcome in order to implement Parallel Commit. This document includes ideas to solve them.
+
+## One-phase Writing Locking Problem
+
+In a Parallel-Commit transaction, once all prewrites is succeeded, we say the transaction is successfully committed. The commit_ts of a transaction `T` should be greater than all reads on `T`'s keys that happens before `T`'s prewrite operations, and less than transactions that starts after `T` committing.
+
+We cannot asynchronously allocate a TSO as transaction `T`'s commit_ts after telling the client that the transaction has been finished, because it's possible that the client runs faster and got a earlier TSO to start its next transaction `T2`, so that in `T2`'s sight the previous transaction `T` didn't commit.
+
+Neither can we tell the client `T` has been committed just after allocating TSO as `T`'s commit_ts. Because if server crashes it will never know what TS it has allocated, and it can not find the proper commit_ts anymore.
+
+Actually, we believe we should persist some information that helps us in finding the commit_ts, and it should have been persisted when the transaction has been "successfully committed". Our current idea is persist `max_read_ts` into the lock that we prewrites, and the final commit_ts will be `max_over_all_keys{max_read_ts}+1`. However it's hard: we need to get the current value of `max_read_ts` before writing down the locks, however new reads may happens between getting `max_read_ts` and successfully writing down the lock.
+
+We may have no perfect solution to this problem, but we have came up with some different ideas that may be possible to solve it.
+
+### Idea 1: Memory Locking 
+
+When we got the max read ts as `M` and then prewriting key `K` with max read ts `M`, we want to prevent other transactions from reading `K` with a ts larger than `M`. The most straightforward way is to make a in-memory lock: A Parallel-Commit prewrite should lock the keys it want to write, before reading the `max_read_ts`, and release the lock after finishing prewrite. A read operation should check the memory lock after updating `max_read_ts` and should be blocked (or return a KeyIsLocked error) when it meets the memory lock. Read operations don't need to actually acquire locks. It can proceed reading if it finds that the keys it wants to read are not locked.

```

Comment:
>  A read operation ... and should be blocked (or return a KeyIsLocked error) when it meets the memory lock.

I just noticed that it should return a KeyIsLocked error directly here, since after the memory lock is released, it will nearly definitely meet the lock we just wrote.

Reply:
If a read found a mem lock, check the mem lock max_read_ts, if read ts is smaller than mem lock's max_read_ts, the read can be served.